include/openmc/boundary_condition.h

```diff
@@ -106,5 +113,49 @@ class RotationalPeriodicBC : public PeriodicBC {
   double angle_;
 };
 
+//==============================================================================
+//! A BC template that modifies the weight of incident particles.
+//
+//! Compatible with ReflectiveBC, WhiteBC and both types of PeriodicBC.
+//==============================================================================
+
+template<typename BaseBC>
+class AlbedoBC : public BaseBC {
+public:
+  AlbedoBC(double albedo) : bc_alb_(albedo) {};
+  AlbedoBC(int i_surf, int j_surf, double albedo)
+    : BaseBC(i_surf, j_surf), bc_alb_(albedo) {};
+  // Default to not modifying particle weights
+  AlbedoBC() : bc_alb_(1.0) {};
+
+  void handle_particle(Particle& p, const Surface& surf) const override
+  {
+    // Save incident particle's weight
+    double initial_wgt = p.wgt();
+    // Treat the lost weight fraction as leakage, similar to VacuumBC.
+    // This ensures the lost weight is tallied properly.
+    p.wgt() *= (1.0 - bc_alb_);
+    p.cross_vacuum_bc(surf);
+    // Reassign the remaining weight and pass particle to BaseBC
+    p.wgt() = initial_wgt * bc_alb_;
+    BaseBC::handle_particle(p, surf);
+  };
```

Comment:
In what sense is it overkill? I suggested this so as to not have irrelevant data on a class like VacuumBC, for example.

Reply:
Overkill in the sense that it deepens the class hierarchy when the existing classes could make do, and it's more to track mentally for someone reading the code. Vacuum BC is the only case where it doesn't make sense; to handle that we can have a default albedo of 1.0 and only invoke the albedo handling if the value is not 1.0.

Reply:
Alright, sounds reasonable to me!

Reply:
@nutcasev15 does that sound acceptable to you?

Reply:
The first revision of this PR was very similar to what you have suggested. However, we determined that the template way was more modular, reduced the fragmentation of code changes across files and used slightly less memory and ran slightly faster.

Ultimately, I think the memory and speed criteria are the most important and I've done a comparison:

**Current `<template>` Implementation:**
No Albedo in XML (Legacy Models):
- Calculation Rate (inactive)       = 76830.7 particles/second
- Calculation Rate (active)         = 78405.9 particles/second

Albedo is 1.0 in XML:
- Calculation Rate (inactive)       = 77180.6 particles/second
- Calculation Rate (active)         = 80555.1 particles/second

**Suggested `BoundaryCondition` Modification Implementation:**
No Albedo in XML (Legacy Models):
- Calculation Rate (inactive)       = 76682.5 particles/second
- Calculation Rate (active)         = 77538.1 particles/second

Albedo is 1.0 in XML:

- Calculation Rate (inactive)       = 76869.9 particles/second
- Calculation Rate (active)         = 78743.8 particles/second

I think we should err on the side of performance because it can have knock on effects on depletion runs and other repetitive eigenvalue calculations. Further, the current implementation ensures that no legacy models will break while making changes to `BoundaryCondition` means users that don't use albedos would still have a slight performance penalty. I vote to go ahead with how things are currently.

The diff of the changes for the `BoundaryCondition` Modification implementation is below:
```
diff --git a/include/openmc/boundary_condition.h b/include/openmc/boundary_condition.h
index fb35b87a8..668f14f0a 100644
--- a/include/openmc/boundary_condition.h
+++ b/include/openmc/boundary_condition.h
@@ -4,7 +4,6 @@
 #include "openmc/hdf5_interface.h"
 #include "openmc/particle.h"
 #include "openmc/position.h"
-#include "openmc/random_lcg.h"
 #include <fmt/core.h>
 
 namespace openmc {
@@ -24,13 +23,39 @@ public:
   //!   to directly modify anything about the particle, but it will do so
   //!   indirectly by calling the particle's appropriate cross_*_bc function.
   //! \param surf The specific surface on the boundary the particle struck.
-  virtual void handle_particle(Particle& p, const Surface& surf) const = 0;
+  virtual void handle_particle(Particle& p, const Surface& surf) const {
+
+    if (!has_alebedo_)
+      return;
+
+    // Save incident particle's weight
+    double initial_wgt = p.wgt();
+    // Treat the lost weight fraction as leakage, similar to VacuumBC.
+    // This ensures the lost weight is tallied properly.
+    p.wgt() *= (1.0 - albedo_);
+    p.cross_vacuum_bc(surf);
+    // Reassign the remaining weight and pass particle to BaseBC
+    p.wgt() = initial_wgt * albedo_;
+  };
 
   //! Return a string classification of this BC.
   virtual std::string type() const = 0;
 
   //! Write BC data such as albedo to hdf5.
-  virtual void to_hdf5(hid_t surf_group) const {};
+  void to_hdf5(hid_t surf_group) const  {
+    if (std::abs(albedo_ - 1.0) <= FP_PRECISION) {
+      write_string(surf_group, "albedo", fmt::format("{}", albedo_), false);
+    }
+  };
+
+  virtual void set_albedo(double albedo) {
+    has_alebedo_ = true;
+    albedo_ = albedo;
+  }
+
+private:
+double albedo_ = 1.0;
+bool has_alebedo_ = false;
 };
 
 //==============================================================================
@@ -113,49 +138,5 @@ protected:
   double angle_;
 };
 
-//==============================================================================
-//! A BC template that modifies the weight of incident particles.
-//
-//! Compatible with ReflectiveBC, WhiteBC and both types of PeriodicBC.
-//==============================================================================
-
-template<typename BaseBC>
-class AlbedoBC : public BaseBC {
-public:
-  AlbedoBC(double albedo) : bc_alb_(albedo) {};
-  AlbedoBC(int i_surf, int j_surf, double albedo)
-    : BaseBC(i_surf, j_surf), bc_alb_(albedo) {};
-  // Default to not modifying particle weights
-  AlbedoBC() : bc_alb_(1.0) {};
-
-  void handle_particle(Particle& p, const Surface& surf) const override
-  {
-    // Save incident particle's weight
-    double initial_wgt = p.wgt();
-    // Treat the lost weight fraction as leakage, similar to VacuumBC.
-    // This ensures the lost weight is tallied properly.
-    p.wgt() *= (1.0 - bc_alb_);
-    p.cross_vacuum_bc(surf);
-    // Reassign the remaining weight and pass particle to BaseBC
-    p.wgt() = initial_wgt * bc_alb_;
-    BaseBC::handle_particle(p, surf);
-  };
-
-  void to_hdf5(hid_t surf_group) const override
-  {
-    if (std::abs(bc_alb_ - 1.0) <= FP_PRECISION) {
-      write_string(surf_group, "albedo", fmt::format("{}", bc_alb_), false);
-    }
-  };
-
-private:
-  double bc_alb_;
-};
-
-typedef AlbedoBC<ReflectiveBC> ReflectiveAlbedoBC;
-typedef AlbedoBC<WhiteBC> WhiteAlbedoBC;
-typedef AlbedoBC<RotationalPeriodicBC> RotationalPeriodicAlbedoBC;
-typedef AlbedoBC<TranslationalPeriodicBC> TranslationalPeriodicAlbedoBC;
-
 } // namespace openmc
 #endif // OPENMC_BOUNDARY_CONDITION_H
diff --git a/src/boundary_condition.cpp b/src/boundary_condition.cpp
index 5b842399e..fa55f4e79 100644
--- a/src/boundary_condition.cpp
+++ b/src/boundary_condition.cpp
@@ -28,6 +28,7 @@ void ReflectiveBC::handle_particle(Particle& p, const Surface& surf) const
   Direction u = surf.reflect(p.r(), p.u(), &p);
   u /= u.norm();
 
+  BoundaryCondition::handle_particle(p, surf);
   p.cross_reflective_bc(surf, u);
 }
 
@@ -40,6 +41,7 @@ void WhiteBC::handle_particle(Particle& p, const Surface& surf) const
   Direction u = surf.diffuse_reflect(p.r(), p.u(), p.current_seed());
   u /= u.norm();
 
+  BoundaryCondition::handle_particle(p, surf);
   p.cross_reflective_bc(surf, u);
 }
 
@@ -130,6 +132,8 @@ void TranslationalPeriodicBC::handle_particle(
       "hitting a surface, but that surface is not recognized by the BC.");
   }
 
+  BoundaryCondition::handle_particle(p, surf);
+
   // Pass the new location and surface to the particle.
   p.cross_periodic_bc(surf, new_r, p.u(), new_surface);
 }
@@ -264,6 +268,8 @@ void RotationalPeriodicBC::handle_particle(
   Direction new_u = {
     cos_theta * u.x - sin_theta * u.y, sin_theta * u.x + cos_theta * u.y, u.z};
 
+  BoundaryCondition::handle_particle(p, surf);
+
   // Pass the new location, direction, and surface to the particle.
   p.cross_periodic_bc(surf, new_r, new_u, new_surface);
 }
diff --git a/src/surface.cpp b/src/surface.cpp
index 8adab6f6a..a0295377e 100644
--- a/src/surface.cpp
+++ b/src/surface.cpp
@@ -77,8 +77,25 @@ Surface::Surface(pugi::xml_node surf_node)
   if (check_for_node(surf_node, "boundary")) {
     std::string surf_bc = get_node_value(surf_node, "boundary", true, true);
 
+    if (surf_bc == "transmission" || surf_bc == "transmit" || surf_bc.empty()) {
+      // Leave the bc_ a nullptr
+    } else if (surf_bc == "vacuum") {
+      bc_ = std::make_shared<VacuumBC>();
+    } else if (surf_bc == "reflective" || surf_bc == "reflect" ||
+               surf_bc == "reflecting") {
+      bc_ = std::make_shared<ReflectiveBC>();
+    } else if (surf_bc == "white") {
+      bc_ = std::make_shared<WhiteBC>();
+    } else if (surf_bc == "periodic") {
+      // Periodic BCs are handled separately
+    } else {
+      fatal_error(fmt::format("Unknown boundary condition \"{}\" specified "
+                              "on surface {}",
+        surf_bc, id_));
+    }
+
     double surf_alb;
-    if (check_for_node(surf_node, "albedo")) {
+    if (check_for_node(surf_node, "albedo") && bc_ != nullptr) {
       surf_alb = std::stod(get_node_value(surf_node, "albedo"));
 
       if (surf_alb < 0.0)
@@ -91,29 +108,8 @@ Surface::Surface(pugi::xml_node surf_node)
                             "Albedos greater than 1 may cause "
                             "unphysical behaviour.",
           id_, surf_alb));
-    }
 
-    if (surf_bc == "transmission" || surf_bc == "transmit" || surf_bc.empty()) {
-      // Leave the bc_ a nullptr
-    } else if (surf_bc == "vacuum") {
-      bc_ = std::make_shared<VacuumBC>();
-    } else if (surf_bc == "reflective" || surf_bc == "reflect" ||
-               surf_bc == "reflecting") {
-      if (check_for_node(surf_node, "albedo"))
-        bc_ = std::make_shared<ReflectiveAlbedoBC>(surf_alb);
-      else
-        bc_ = std::make_shared<ReflectiveBC>();
-    } else if (surf_bc == "white") {
-      if (check_for_node(surf_node, "albedo"))
-        bc_ = std::make_shared<WhiteAlbedoBC>(surf_alb);
-      else
-        bc_ = std::make_shared<WhiteBC>();
-    } else if (surf_bc == "periodic") {
-      // Periodic BCs are handled separately
-    } else {
-      fatal_error(fmt::format("Unknown boundary condition \"{}\" specified "
-                              "on surface {}",
-        surf_bc, id_));
+      bc_->set_albedo(surf_alb);
     }
   }
 }
@@ -1334,31 +1330,19 @@ void read_surfaces(pugi::xml_node node)
     // condition.  Otherwise, it is a rotational periodic BC. If albedo data
     // is present, instantiate appropriate BC type.
     if (std::abs(1.0 - dot_prod) < FP_PRECISION) {
-      if (albedo_map.count(surf1.id_)) {
-        surf1.bc_ = std::make_shared<TranslationalPeriodicAlbedoBC>(
-          i_surf, j_surf, albedo_map[surf1.id_]);
-      } else {
-        surf1.bc_ = std::make_shared<TranslationalPeriodicBC>(i_surf, j_surf);
-      }
-      if (albedo_map.count(surf2.id_)) {
-        surf2.bc_ = std::make_shared<TranslationalPeriodicAlbedoBC>(
-          i_surf, j_surf, albedo_map[surf2.id_]);
-      } else {
-        surf2.bc_ = std::make_shared<TranslationalPeriodicBC>(i_surf, j_surf);
-      }
+      surf1.bc_ = std::make_shared<TranslationalPeriodicBC>(i_surf, j_surf);
+      surf2.bc_ = std::make_shared<TranslationalPeriodicBC>(i_surf, j_surf);
     } else {
-      if (albedo_map.count(surf1.id_)) {
-        surf1.bc_ = std::make_shared<RotationalPeriodicAlbedoBC>(
-          i_surf, j_surf, albedo_map[surf1.id_]);
-      } else {
-        surf1.bc_ = std::make_shared<RotationalPeriodicBC>(i_surf, j_surf);
-      }
-      if (albedo_map.count(surf2.id_)) {
-        surf2.bc_ = std::make_shared<RotationalPeriodicAlbedoBC>(
-          i_surf, j_surf, albedo_map[surf2.id_]);
-      } else {
-        surf2.bc_ = std::make_shared<RotationalPeriodicBC>(i_surf, j_surf);
-      }
+      surf1.bc_ = std::make_shared<RotationalPeriodicBC>(i_surf, j_surf);
+      surf2.bc_ = std::make_shared<RotationalPeriodicBC>(i_surf, j_surf);
+    }
+
+    if (albedo_map.count(surf1.id_)) {
+        surf1.bc_->set_albedo(albedo_map[surf1.id_]);
+    }
+
+    if (albedo_map.count(surf2.id_)) {
+        surf2.bc_->set_albedo(albedo_map[surf2.id_]);
     }
   }
 }
```

Reply:
Thanks for the detailed analysis @nutcasev15. What is the test problem you are running here?

Reply:
Its the regression test that's included in the PR. I think it's a good representation of models where there is a large amount of interaction with the boundary conditions.

Reply:
I tried out both implementations on my machine and I'm seeing very little difference in performance, at most 1–2%, although even then it's hard to get solid numbers because the normal machine jitter tends to be more than that. Keeping in mind that this is on a very contrived problem with particles bouncing back and forth off boundaries, for a practical problem we will see effectively 0 difference in performance.

Regarding memory, you can get away with storing just the `albedo_` member (whether a BC "has" an albedo can be checked by comparing to 1.0), which means that you have at most an extra 8 bytes per boundary condition. I've never seen a model with more than six explicitly specified BCs, so the memory cost is at most 48 bytes.

One worry I have with the template approach is its impact for a future GPU implementation. We have to explicitly instantiate each concrete class, which results in a bunch of extra code that has to be compiled that is most likely not used (on my system, the templated approach results in `libopenmc.so` being 200 kB larger). On CPU this is not a big deal, but on GPU you pay a price to have more code within a kernel because it affects register usage and hence occupancy.

So, overall I don't think the potentially small performance benefit on the test problem is worth the extra complexity and possible negative impact down the line for GPU porting purposes.

Reply:
I fully understand and agree with you that the performance gain minimal. PR updated @paulromano! I hope this is merge ready?

Reply:
Good point Paul on the GPU stuff, although I suspect the compiler would be able to eliminate the extra code if we went with a tagged union approach to replace polymorphism for this hypothetical nicely written GPU version. If the BC classes are just functors, it could inline the code in the switch/case statement that gets generated by a variant.

Regardless, I think putting the albedo on the BC makes a lot of sense. Previously it was on each surface... somehow I didn't think about this possibility. It can be tempting to jump to one of these fancy "design patterns" just to use some fun C++ features. 😄